kodein-core/src/main/kotlin/com/github/salomonbrys/kodein/types.kt

```diff
@@ -69,16 +69,21 @@ internal abstract class JVMTypeToken<T> : TypeToken<T> {
 
     override fun equals(other: Any?): Boolean {
         if (this === other) return true
+        if (other is CompositeTypeToken<*>) {
+            return fullDispString().equals(this.fullDispString())
+        }
         if (other !is JVMTypeToken<*>) return false
 
         if (type() != other.type()) return false
 
         return true
     }
 
-    override fun hashCode(): Int {
-        return type().hashCode()
-    }
+    override fun hashCode() = fullDispString().hashCode()
```

Comment:
for CMap hash find use fullDispString's hashcode.
So, generic&lt;A&lt;B&gt;&gt;().hashCode() == erasedComp1&lt;A&lt;B&gt;&gt;(erased(), erased&lt;B&gt;()).hashCode()

Reply:
This is very interesting !